### ISSUE TITLE
Paste Images into Markdown document

### ISSUE THREAD

**Issue by longbuivan** [zed-industries/zed#18112]
### Summary

Support pasting images from clipboard into Markdown and other documents being edited. 

### Description

Functionality similar to GitHub when editing a Markdown document, whereby when you paste image data from the clipboard it does the following:

1. Creates an image file from the pasted clipboard data in a folder adjacent to the file you are currently editing. (e.g. something for something.md/something.py)
2. The relative path of that new file is pasted into the buffer, `./something/something001.png`.
3. When editing a Markdown file instead of `./something/something001.png` you'd get `![](./something/something001.png)`

Optionally include a setting with a filename template to configure the folder/file name where images should be pasted (e.g. `./${basename}_${counter}`, `.images/pasted_image_${counter}`, etc).

Extension prior art:
- Sublime: https://github.com/robinchenyu/imagepaste/
- VSCode: https://marketplace.visualstudio.com/items?itemName=mushan.vscode-paste-image
- Neovim: https://github.com/HakonHarnes/img-clip.nvim

**Comment by Rhelvetican** [comment]
This might be a little hard because different languages have different ways to represent an image. 

Also pardon me for asking but did you write the description by AI?

**Comment by intrd** [comment]
In Sublime, there's a plugin that solves this in a simple way:
- It saves the image from the clipboard to the current folder of the .md file I’m editing
- Then inserts its path ![](images/image1.png) at the cursor.

![image](https://github.com/user-attachments/assets/1319d732-b0b8-4dca-8a87-ec84c4b763eb)

The plugin: https://github.com/robinchenyu/imagepaste/

How to do this on Zed? any pointing?

**Comment by Rhelvetican** [comment]
@intrd By referencing it (in not .md and similar languages), you mean pasting the (presumably relative) path to the image?

For Rust we can use `include_bytes!(<path>)`.

**Comment by notpeter** [comment]
@longbuivan If I am understanding this request correctly, you would like to have functionality similar to GitHub when editing a Markdown document, whereby when you paste image data from the clipboard it does the following:

1. Creates an image file from the pasted clipboard data in a folder adjacent to the file you are currently editing.  The folder/filename would be based on the filename of the currently edited file (e.g. `ImagePaste` for `ImagePaste.py`)  
2. The relative path of that new file is pasted into the buffer, `./ImagePaste/ImagePaste001.png`.
3. When editing a Markdown file instead of `./ImagePaste/ImagePaste001.png` you'd get `![](./ImagePaste/ImagePaste001.png)`

Nifty.

**Comment by intrd** [comment]
Yes, it's exactly that, @notpeter. I think it's a good idea for an extension, not a Zed built-in feature.

**Comment by hedefalk** [comment]
This is the one I've been using on vscode https://marketplace.visualstudio.com/items?itemName=mushan.vscode-paste-image

**Comment by noateden** [comment]
This feature is absolutely helpful, here is step by step how to do it:
- Find a image on web browser
- Right-lick, copy the image (any kind of format: png, svg, jpeg, ...)
- Go to editor (Zed, VsCode, ...) where you want to save the image
- Just paste from clipboard, editor should save the image in PNG format.

It's working on vscode without any extensions & plugins.

**Comment by longbuivan** [comment]
> This might be a little hard because different languages have different ways to represent an image. 
> 
> 
> 
> Also pardon me for asking but did you write the description by AI?

I think that would be enveloped by extension... sorry it is not my expertise. 

NW. Honestly, I was writing by myself.

**Comment by longbuivan** [comment]
Thanks everyone for your clarification.

Since Zed is now my main coding editor in AI-power programming. The workflow for copy and paste to add image into Markdown file is seemly critical my case and I do believe someone like me.

In this moment, I work around with save the image into the "pinned item" in Finder(MacOS file explorer). It works by then.

But it would be significant improvement if we natively support within Zed.

Best,

**Comment by Moelf** [comment]
for reference, an example plug-in from neovim world is: https://github.com/HakonHarnes/img-clip.nvim

**Comment by aizigao** [comment]
I tried using task to solve this problem on macOS. I created a script.

My notes directory is structured like this

```sh
.
├── attachments
│   ├── test__2022-04-13-09-53-24.png
│   ├── test__2022-06-17-13-33-46.png
├── notes
│   └── test.md
├── readme.md
├── scripts
│   ├── clean_useless_attachments.js
│   └── sync.js
└── zed-image-paste.sh

```

add a task

```json
 {
        "label": "zed Paste Img",
        "command": "/Users/aizigao/MyWorkSpace/notesssss/zed-image-paste.sh",
        "env": {
            "IMG_SAVE_PATH": "attachments"
        },
        "use_new_terminal": true,
        "allow_concurrent_runs": true,
        "reveal": "never",
        "hide": "always",
        "shell": "system"
    }
```

zed-image-paste.sh
```sh
#!/usr/bin/env bash
set -e
if [[ "$ZED_FILENAME" != *.md ]]; then
    echo "Not a Markdown file"
    exit -1
fi

ZED_FILENAME_WITHOUT_EXTENSION=$(basename "$ZED_FILENAME" | sed 's/\.[^.]*$//')

PASTE_IMAGE_SAVE_PATH=$ZED_WORKTREE_ROOT/$IMG_SAVE_PATH


# Create the target directory
mkdir -p $PASTE_IMAGE_SAVE_PATH


# Get a timestamp
timestamp=$(date +"%Y-%m-%d-%H-%M-%S")
filename="$ZED_FILENAME_WITHOUT_EXTENSION-$timestamp.png"
filepath="$PASTE_IMAGE_SAVE_PATH/$filename"

# Use AppleScript to get the clipboard image and save it
osascript <<EOF
try
    set theImage to the clipboard as «class PNGf»
    set theFile to POSIX file "$filepath"
    set outFile to open for access theFile with write permission
    set eof outFile to 0
    write theImage to outFile
    close access outFile
on error
    return "no_image"
end try
EOF



# Check whether the image was generated successfully
if [[ -f "$filepath" ]]; then
    echo gen img $filepath
    RELATIVE_PATH=$(python3 -c "import os; print(os.path.relpath('$PASTE_IMAGE_SAVE_PATH', '$ZED_DIRNAME'))")
    # Generate Markdown and copy it to the clipboard
    echo "![${filename}]($RELATIVE_PATH/${filename})" | pbcopy
    echo ok
else
    echo "no image buffer in clipboard"
    exit -100
fi


```
run 
```
chmod +x zed-image-paste.sh
```

keymap
```json
  {
        "context": "Workspace",
        "bindings": {
            "alt-ctrl-p": ["task::Spawn", { "task_name": "zed Paste Img" }]
        }
    }
```

I usually copy an image and then run image-paste.
It will write the image to the clipboard, and then I just manually paste it (I do know how to call editor:Paste action in task).


https://github.com/user-attachments/assets/c1aad466-66dc-4806-8d80-b5b6f8ac10f8

**Comment by meltingscales** [comment]
Does a plugin exist that can do this?

**Comment by KeeTraxx** [comment]
Since I write a lot of markdown files, but I'm annoyed that I have to switch to vscode to paste images from clipboard, I decided to add zed native support for pasting images into markdown files in https://github.com/zed-industries/zed/pull/58588

I tested it on linux on my local machine and it seems to work quite fine.

I would welcome if this would be merged, as zed extensions cannot cover this feature.
Also the "workaround" through zed tasks does not really make me happy.

Works with multiple cursors too.

Didn't test, but I don't think it will cause issues:

- Windows / OSX
- Pasting in vim or other modes

**Comment by 10-24** [comment]
Typst?
